File: frontend/src/api/knowledge.ts
import type { DataResponse, ListResponse } from "../types";
import { http } from "./client";

export interface KnowledgeDoc {
  id: string;
  code: string;
  slug: string;
  title: string;
  doc_type: string;
  category_id: string | null;
  category_name: string | null;
  summary: string | null;
  status: string;
  version: number;
  tags: string[];
  is_pinned: boolean;
  view_count: number;
  author_id: string | null;
  approver_id: string | null;
  updated_at: string | null;
  created_at: string | null;
  // present on the detail endpoint
  content?: string;
  keywords?: string | null;
  relationships?: Relationship[];
  references?: Reference[];
}

export interface Category {
  id: string;
  code: string;
  name: string;
  description: string | null;
  parent_id: string | null;
  position: number;
  document_count: number | null;
}

export interface Relationship {
  id: string;
  source_document_id: string;
  source_title: string | null;
  target_document_id: string;
  target_title: string | null;
  relationship_type: string;
  note: string | null;
}

export interface Reference {
  id: string;
  document_id: string;
  entity_type: string;
  entity_id: string | null;
  label: string | null;
}

export interface DocVersion {
  id: string;
  document_id: string;
  version: number;
  title: string;
  change_summary: string | null;
  status: string;
  created_at: string | null;
}

export interface Review {
  id: string;
  document_id: string;
  reviewer_name: string | null;
  decision: string;
  comment: string | null;
  reviewed_at: string | null;
}

export interface Collection {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  owner_id: string | null;
  document_count: number;
  documents?: { document_id: string; title: string | null; position: number }[];
}

export interface Bookmark {
  id: string;
  document_id: string;
  document_title: string | null;
  document_code: string | null;
  note: string | null;
  created_at: string | null;
}

export interface ADR {
  id: string;
  code: string;
  title: string;
  status: string;
  context: string | null;
  decision: string | null;
  consequences: string | null;
  document_id: string | null;
  decided_at: string | null;
  created_at: string | null;
}

export interface GraphData {
  nodes: {
    id: string;
    code: string;
    title: string;
    doc_type: string;
    category_id: string | null;
  }[];
  edges: { id: string; source: string; target: string; type: string }[];
}

export interface DocBrief {
  id: string;
  code: string;
  title: string;
  doc_type: string;
  summary: string | null;
}

export interface RetrievalBundle {
  relevant_documents: DocBrief[];
  relevant_sop: DocBrief[];
  relevant_adr: DocBrief[];
  relevant_standards: DocBrief[];
  relevant_decisions: DocBrief[];
  relevant_templates: DocBrief[];
  relevant_references: { document_id: string; entity_type: string; label: string | null }[];
}

export interface KnowledgeFounderDash {
  documents: number;
  categories: number;
  pending_reviews: number;
  approved_documents: number;
  knowledge_health: string;
  approved_coverage: number;
  recent_knowledge: KnowledgeDoc[];
  most_used: KnowledgeDoc[];
  statistics: {
    total_documents: number;
    total_categories: number;
    total_adrs: number;
    total_views: number;
    retrievals: number;
    by_status: Record<string, number>;
    by_type: Record<string, number>;
  };
}

export interface KnowledgeAIDash {
  suggested_knowledge: DocBrief[];
  recent_knowledge: KnowledgeDoc[];
  architecture_references: DocBrief[];
  coding_standards: DocBrief[];
  sop_recommendations: DocBrief[];
  organization_memory: DocBrief[];
  related_documents: DocBrief[];
}

export interface DocumentInput {
  title: string;
  doc_type: string;
  content?: string;
  summary?: string | null;
  category_id?: string | null;
  keywords?: string | null;
  tags?: string[];
}

const qs = (params: Record<string, string | undefined>): string => {
  const entries = Object.entries(params).filter(([, v]) => v);
  return entries.length
    ? `?${entries.map(([k, v]) => `${k}=${encodeURIComponent(v as string)}`).join("&")}`
    : "";
};

export const knowledgeApi = {
  // documents
  documents: (filters: { category_id?: string; doc_type?: string; status?: string } = {}) =>
    http.get<ListResponse<KnowledgeDoc>>(`/knowledge/documents${qs(filters)}`),
  document: (id: string) => http.get<DataResponse<KnowledgeDoc>>(`/knowledge/documents/${id}`),
  createDocument: (input: DocumentInput) =>
    http.post<DataResponse<KnowledgeDoc>>("/knowledge/documents", input),
  updateDocument: (id: string, input: Partial<DocumentInput> & { change_summary?: string }) =>
    http.patch<DataResponse<KnowledgeDoc>>(`/knowledge/documents/${id}`, input),
  versions: (id: string) =>
    http.get<ListResponse<DocVersion>>(`/knowledge/documents/${id}/versions`),
  restore: (id: string, version: number) =>
    http.post<DataResponse<KnowledgeDoc>>(
      `/knowledge/documents/${id}/versions/${version}/restore`,
      {},
    ),
  related: (id: string) =>
    http.get<ListResponse<KnowledgeDoc>>(`/knowledge/documents/${id}/related`),
  submit: (id: string) =>
    http.post<DataResponse<KnowledgeDoc>>(`/knowledge/documents/${id}/submit`, {}),
  reviews: (id: string) => http.get<ListResponse<Review>>(`/knowledge/documents/${id}/reviews`),
  review: (id: string, decision: string, comment?: string) =>
    http.post<DataResponse<Review>>(`/knowledge/documents/${id}/review`, { decision, comment }),
  addReference: (id: string, entity_type: string, label?: string) =>
    http.post<DataResponse<Reference>>(`/knowledge/documents/${id}/references`, {
      entity_type,
      label,
    }),
  // categories / tags
  categories: () => http.get<ListResponse<Category>>("/knowledge/categories"),
  // search / retrieval
  search: (filters: {
    q?: string;
    category_id?: string;
    doc_type?: string;
    tag?: string;
    status?: string;
  }) => http.get<ListResponse<KnowledgeDoc>>(`/knowledge/search${qs(filters)}`),
  retrieve: (keywords?: string) =>
    http.get<DataResponse<RetrievalBundle>>(`/knowledge/retrieve${qs({ keywords })}`),
  // graph
  graph: () => http.get<DataResponse<GraphData>>("/knowledge/graph"),
  createRelationship: (source: string, target: string, type: string, note?: string) =>
    http.post<DataResponse<Relationship>>("/knowledge/relationships", {
      source_document_id: source,
      target_document_id: target,
      relationship_type: type,
      note,
    }),
  // dashboards
  founderDashboard: () =>
    http.get<DataResponse<KnowledgeFounderDash>>("/knowledge/founder-dashboard"),
  aiDashboard: (keywords?: string) =>
    http.get<DataResponse<KnowledgeAIDash>>(`/knowledge/ai-dashboard${qs({ keywords })}`),
  // reviews
  pendingReviews: () => http.get<ListResponse<KnowledgeDoc>>("/knowledge/reviews/pending"),
  // adrs
  adrs: () => http.get<ListResponse<ADR>>("/knowledge/adrs"),
  createAdr: (input: {
    title: string;
    context?: string;
    decision?: string;
    consequences?: string;
  }) => http.post<DataResponse<ADR>>("/knowledge/adrs", input),
  setAdrStatus: (id: string, status: string) =>
    http.patch<DataResponse<ADR>>(`/knowledge/adrs/${id}/status`, { status }),
  // bookmarks
  bookmarks: () => http.get<ListResponse<Bookmark>>("/knowledge/bookmarks"),
  addBookmark: (document_id: string, note?: string) =>
    http.post<DataResponse<Bookmark>>("/knowledge/bookmarks", { document_id, note }),
  removeBookmark: (document_id: string) => http.del(`/knowledge/bookmarks/${document_id}`),
  // collections
  collections: () => http.get<ListResponse<Collection>>("/knowledge/collections"),
  collection: (id: string) => http.get<DataResponse<Collection>>(`/knowledge/collections/${id}`),
  createCollection: (name: string, description?: string) =>
    http.post<DataResponse<Collection>>("/knowledge/collections", { name, description }),
  addToCollection: (collectionId: string, document_id: string) =>
    http.post<DataResponse<Collection>>(`/knowledge/collections/${collectionId}/documents`, {
      document_id,
    }),
};

export const DOC_TYPES = [
  "architecture",
  "adr",
  "sop",
  "specification",
  "api",
  "design",
  "meeting_notes",
  "research",
  "reference",
  "coding_standard",
  "security_standard",
  "deployment_guide",
  "troubleshooting_guide",
  "project_documentation",
  "lessons_learned",
  "policy",
  "template",
];

export const docTypeLabel = (t: string): string =>
  t.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());

File: backend/app/api/v1/knowledge.py
"""Organizational Knowledge Engine endpoints (Sprint 10).

Documents (CRUD + versions), categories, tags, the knowledge graph
(relationships + references), search, AI retrieval, reviews/approvals,
bookmarks, collections, ADRs, and the founder/AI dashboards.

RBAC: read → all authenticated roles; write (author/edit) → knowledge:write;
approvals → knowledge:approve.
"""

import uuid

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from app.api.deps import (
    CurrentUser,
    get_adr_service,
    get_approval_service,
    get_bookmark_service,
    get_collection_service,
    get_current_user,
    get_knowledge_analytics_service,
    get_knowledge_graph_service,
    get_knowledge_service,
    get_reference_service,
    get_retrieval_service,
    get_search_service,
    get_version_service,
    require_permission,
)
from app.domain.knowledge_enums import (
    ADRStatus,
    DocumentType,
    ReferenceEntityType,
    RelationshipType,
    ReviewDecision,
)
from app.domain.roles import Permission

router = APIRouter(prefix="/knowledge", tags=["knowledge"])
_read = Depends(require_permission(Permission.KNOWLEDGE_READ))
_write = Depends(require_permission(Permission.KNOWLEDGE_WRITE))
_approve = Depends(require_permission(Permission.KNOWLEDGE_APPROVE))


# --- payloads -------------------------------------------------------------


class DocumentIn(BaseModel):
    title: str = Field(min_length=2, max_length=300)
    doc_type: DocumentType
    content: str = ""
    summary: str | None = None
    category_id: uuid.UUID | None = None
    keywords: str | None = None
    owner_ai_employee_id: uuid.UUID | None = None
    tags: list[str] | None = None


class DocumentUpdateIn(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=300)
    content: str | None = None
    summary: str | None = None
    category_id: uuid.UUID | None = None
    keywords: str | None = None
    tags: list[str] | None = None
    change_summary: str | None = None


class CategoryIn(BaseModel):
    code: str = Field(min_length=2, max_length=60)
    name: str = Field(min_length=2, max_length=160)
    description: str | None = None
    position: int = 0


class RelationshipIn(BaseModel):
    source_document_id: uuid.UUID
    target_document_id: uuid.UUID
    relationship_type: RelationshipType
    note: str | None = None


class ReferenceIn(BaseModel):
    entity_type: ReferenceEntityType
    entity_id: uuid.UUID | None = None
    label: str | None = None


class ReviewIn(BaseModel):
    decision: ReviewDecision
    comment: str | None = None


class BookmarkIn(BaseModel):
    document_id: uuid.UUID
    note: str | None = None


class CollectionIn(BaseModel):
    name: str = Field(min_length=2, max_length=200)
    description: str | None = None


class CollectionDocumentIn(BaseModel):
    document_id: uuid.UUID


class ADRIn(BaseModel):
    title: str = Field(min_length=2, max_length=300)
    context: str | None = None
    decision: str | None = None
    consequences: str | None = None
    status: ADRStatus = ADRStatus.PROPOSED
    document_id: uuid.UUID | None = None
    project_id: uuid.UUID | None = None


class ADRStatusIn(BaseModel):
    status: ADRStatus


# --- categories & tags ----------------------------------------------------


@router.get("/categories", dependencies=[_read])
def list_categories(service=Depends(get_knowledge_service)) -> dict:
    items = service.categories.with_counts()
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/categories", dependencies=[_write])
def create_category(payload: CategoryIn, service=Depends(get_knowledge_service)) -> dict:
    c = service.categories.create(payload.code, payload.name, payload.description, payload.position)
    return {"data": service.categories.serialize(c, 0)}


@router.get("/tags", dependencies=[_read])
def list_tags(service=Depends(get_knowledge_service)) -> dict:
    items = [service.tags.serialize(t) for t in service.tags.list_tags()]
    return {"data": items, "meta": {"total": len(items)}}


# --- search & retrieval ---------------------------------------------------


@router.get("/search", dependencies=[_read])
def search(
    q: str | None = Query(default=None),
    category_id: uuid.UUID | None = Query(default=None),
    doc_type: str | None = Query(default=None),
    tag: str | None = Query(default=None),
    status: str | None = Query(default=None),
    knowledge=Depends(get_knowledge_service),
    search_service=Depends(get_search_service),
) -> dict:
    docs = search_service.search(
        q, category_id=category_id, doc_type=doc_type, tag=tag, status=status
    )
    items = [knowledge.serialize(d) for d in docs]
    return {"data": items, "meta": {"total": len(items)}}


@router.get("/retrieve", dependencies=[_read])
def retrieve(
    keywords: str | None = Query(default=None),
    ai_employee_id: uuid.UUID | None = Query(default=None),
    service=Depends(get_retrieval_service),
) -> dict:
    return {"data": service.retrieve_for(keywords=keywords, ai_employee_id=ai_employee_id)}


# --- dashboards -----------------------------------------------------------


@router.get("/founder-dashboard", dependencies=[_read])
def founder_dashboard(service=Depends(get_knowledge_analytics_service)) -> dict:
    return {"data": service.founder_dashboard()}


@router.get("/ai-dashboard", dependencies=[_read])
def ai_dashboard(
    keywords: str | None = Query(default=None),
    service=Depends(get_knowledge_analytics_service),
) -> dict:
    return {"data": service.ai_dashboard(keywords)}


@router.get("/analytics", dependencies=[_read])
def analytics(service=Depends(get_knowledge_analytics_service)) -> dict:
    return {"data": service.statistics()}


# --- knowledge graph ------------------------------------------------------


@router.get("/graph", dependencies=[_read])
def graph(service=Depends(get_knowledge_graph_service)) -> dict:
    return {"data": service.graph()}


@router.post("/relationships", dependencies=[_write])
def create_relationship(
    payload: RelationshipIn, service=Depends(get_knowledge_graph_service)
) -> dict:
    rel = service.link(
        payload.source_document_id,
        payload.target_document_id,
        payload.relationship_type,
        payload.note,
    )
    return {"data": service.serialize(rel)}


@router.delete("/relationships/{relationship_id}", dependencies=[_write])
def delete_relationship(
    relationship_id: uuid.UUID, service=Depends(get_knowledge_graph_service)
) -> dict:
    service.unlink(relationship_id)
    return {"data": {"deleted": str(relationship_id)}}


# --- reviews (pending list at the collection level) -----------------------


@router.get("/reviews/pending", dependencies=[_read])
def pending_reviews(service=Depends(get_approval_service)) -> dict:
    from app.services.knowledge import KnowledgeService

    ks = KnowledgeService(service.db)
    items = [ks.serialize(d) for d in service.pending_reviews()]
    return {"data": items, "meta": {"total": len(items)}}


# --- ADRs -----------------------------------------------------------------


@router.get("/adrs", dependencies=[_read])
def list_adrs(service=Depends(get_adr_service)) -> dict:
    items = [service.serialize(a) for a in service.list_adrs()]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/adrs", dependencies=[_write])
def create_adr(payload: ADRIn, service=Depends(get_adr_service)) -> dict:
    adr = service.create(
        title=payload.title,
        context=payload.context,
        decision=payload.decision,
        consequences=payload.consequences,
        status=payload.status,
        document_id=payload.document_id,
        project_id=payload.project_id,
    )
    return {"data": service.serialize(adr)}


@router.patch("/adrs/{adr_id}/status", dependencies=[_write])
def set_adr_status(
    adr_id: uuid.UUID, payload: ADRStatusIn, service=Depends(get_adr_service)
) -> dict:
    return {"data": service.serialize(service.set_status(adr_id, payload.status))}


# --- bookmarks ------------------------------------------------------------


@router.get("/bookmarks", dependencies=[_read])
def list_bookmarks(
    user: CurrentUser = Depends(get_current_user), service=Depends(get_bookmark_service)
) -> dict:
    items = [service.serialize(b) for b in service.list_for_user(user.id)]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/bookmarks", dependencies=[_read])
def add_bookmark(
    payload: BookmarkIn,
    user: CurrentUser = Depends(get_current_user),
    service=Depends(get_bookmark_service),
) -> dict:
    bm = service.add(user.id, payload.document_id, payload.note)
    return {"data": service.serialize(bm)}


@router.delete("/bookmarks/{document_id}", dependencies=[_read])
def remove_bookmark(
    document_id: uuid.UUID,
    user: CurrentUser = Depends(get_current_user),
    service=Depends(get_bookmark_service),
) -> dict:
    service.remove(user.id, document_id)
    return {"data": {"removed": str(document_id)}}


# --- collections ----------------------------------------------------------


@router.get("/collections", dependencies=[_read])
def list_collections(service=Depends(get_collection_service)) -> dict:
    items = [service.serialize(c) for c in service.list_collections()]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/collections", dependencies=[_write])
def create_collection(
    payload: CollectionIn,
    user: CurrentUser = Depends(get_current_user),
    service=Depends(get_collection_service),
) -> dict:
    c = service.create(payload.name, description=payload.description, owner_id=user.id)
    return {"data": service.serialize(c)}


@router.get("/collections/{collection_id}", dependencies=[_read])
def get_collection(collection_id: uuid.UUID, service=Depends(get_collection_service)) -> dict:
    return {"data": service.serialize(service.get(collection_id), with_documents=True)}


@router.post("/collections/{collection_id}/documents", dependencies=[_write])
def add_to_collection(
    collection_id: uuid.UUID,
    payload: CollectionDocumentIn,
    service=Depends(get_collection_service),
) -> dict:
    c = service.add_document(collection_id, payload.document_id)
    return {"data": service.serialize(c, with_documents=True)}


@router.delete("/collections/{collection_id}/documents/{document_id}", dependencies=[_write])
def remove_from_collection(
    collection_id: uuid.UUID, document_id: uuid.UUID, service=Depends(get_collection_service)
) -> dict:
    service.remove_document(collection_id, document_id)
    return {"data": service.serialize(service.get(collection_id), with_documents=True)}


# --- documents ------------------------------------------------------------


@router.get("/documents", dependencies=[_read])
def list_documents(
    category_id: uuid.UUID | None = Query(default=None),
    doc_type: str | None = Query(default=None),
    status: str | None = Query(default=None),
    service=Depends(get_knowledge_service),
) -> dict:
    docs = service.list_documents(category_id=category_id, doc_type=doc_type, status=status)
    items = [service.serialize(d) for d in docs]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/documents", dependencies=[_write])
def create_document(
    payload: DocumentIn,
    user: CurrentUser = Depends(get_current_user),
    service=Depends(get_knowledge_service),
) -> dict:
    doc = service.create(
        title=payload.title,
        doc_type=payload.doc_type,
        content=payload.content,
        summary=payload.summary,
        category_id=payload.category_id,
        keywords=payload.keywords,
        author_id=user.id,
        owner_ai_employee_id=payload.owner_ai_employee_id,
        tags=payload.tags,
    )
    return {"data": service.serialize(doc, full=True)}


@router.get("/documents/{document_id}", dependencies=[_read])
def get_document(
    document_id: uuid.UUID,
    graph=Depends(get_knowledge_graph_service),
    references=Depends(get_reference_service),
    service=Depends(get_knowledge_service),
) -> dict:
    doc = service.view(document_id)
    data = service.serialize(doc, full=True)
    data["relationships"] = [graph.serialize(r) for r in graph.for_document(document_id)]
    data["references"] = [references.serialize(r) for r in references.for_document(document_id)]
    return {"data": data}


@router.patch("/documents/{document_id}", dependencies=[_write])
def update_document(
    document_id: uuid.UUID, payload: DocumentUpdateIn, service=Depends(get_knowledge_service)
) -> dict:
    doc = service.update(
        document_id,
        title=payload.title,
        content=payload.content,
        summary=payload.summary,
        category_id=payload.category_id,
        keywords=payload.keywords,
        tags=payload.tags,
        change_summary=payload.change_summary,
    )
    return {"data": service.serialize(doc, full=True)}


@router.get("/documents/{document_id}/versions", dependencies=[_read])
def document_versions(document_id: uuid.UUID, service=Depends(get_version_service)) -> dict:
    items = [service.serialize(v) for v in service.history(document_id)]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/documents/{document_id}/versions/{version}/restore", dependencies=[_write])
def restore_version(
    document_id: uuid.UUID,
    version: int,
    version_service=Depends(get_version_service),
    knowledge=Depends(get_knowledge_service),
) -> dict:
    doc = version_service.restore(document_id, version)
    return {"data": knowledge.serialize(doc, full=True)}


@router.get("/documents/{document_id}/related", dependencies=[_read])
def related_documents(
    document_id: uuid.UUID,
    graph=Depends(get_knowledge_graph_service),
    knowledge=Depends(get_knowledge_service),
) -> dict:
    items = [knowledge.serialize(d) for d in graph.related_documents(document_id)]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/documents/{document_id}/references", dependencies=[_write])
def add_reference(
    document_id: uuid.UUID, payload: ReferenceIn, service=Depends(get_reference_service)
) -> dict:
    ref = service.add(document_id, payload.entity_type, payload.entity_id, payload.label)
    return {"data": service.serialize(ref)}


@router.post("/documents/{document_id}/submit", dependencies=[_write])
def submit_for_review(document_id: uuid.UUID, service=Depends(get_approval_service)) -> dict:
    from app.services.knowledge import KnowledgeService

    doc = service.submit_for_review(document_id)
    return {"data": KnowledgeService(service.db).serialize(doc)}


@router.get("/documents/{document_id}/reviews", dependencies=[_read])
def document_reviews(document_id: uuid.UUID, service=Depends(get_approval_service)) -> dict:
    items = [service.serialize(r) for r in service.reviews_for(document_id)]
    return {"data": items, "meta": {"total": len(items)}}


@router.post("/documents/{document_id}/review", dependencies=[_approve])
def review_document(
    document_id: uuid.UUID,
    payload: ReviewIn,
    user: CurrentUser = Depends(get_current_user),
    service=Depends(get_approval_service),
) -> dict:
    review = service.review(
        document_id, payload.decision, reviewer_id=user.id, comment=payload.comment
    )
    return {"data": service.serialize(review)}
